rir/src/ir/BC.cpp

```diff
@@ -156,6 +157,282 @@ SEXP BC::immediateConst() const {
+            break;
+        case Opcode::guard_fun_:
+            i.guard_fun_args.name = Pool::insert(ReadItem(refTable, inp));
+            i.guard_fun_args.expected = Pool::insert(ReadItem(refTable, inp));
+            i.guard_fun_args.id = InInteger(inp);
+            break;
+        case Opcode::mk_stub_env_:
+        case Opcode::mk_env_: {
+            InBytes(inp, code + 1, sizeof(MkEnvFixedArgs));
+            BC::PoolIdx* names =
+                (BC::PoolIdx*)(code + 1 + sizeof(MkEnvFixedArgs));
+            for (unsigned j = 0; j < i.mkEnvFixedArgs.nargs; j++)
+                names[j] = Pool::insert(ReadItem(refTable, inp));
+            break;
+        }
+        case Opcode::call_:
+        case Opcode::call_implicit_:
+        case Opcode::named_call_:
+        case Opcode::named_call_implicit_: {
+            i.callFixedArgs.nargs = InInteger(inp);
+            i.callFixedArgs.ast = Pool::insert(ReadItem(refTable, inp));
+            InBytes(inp, &i.callFixedArgs.given, sizeof(Assumptions));
+            Opcode* c = code + 1 + sizeof(CallFixedArgs);
+            // Read implicit promise argument offsets
+            if (*code == Opcode::call_implicit_ ||
+                *code == Opcode::named_call_implicit_) {
+                ArgIdx* immArgs = (ArgIdx*)c;
+                for (size_t j = 0; j < i.callFixedArgs.nargs; j++)
+                    immArgs[j] = InInteger(inp);
+                c += i.callFixedArgs.nargs * sizeof(ArgIdx);
+            }
+            // Read named arguments
+            if (*code == Opcode::named_call_ ||
+                *code == Opcode::named_call_implicit_) {
+                PoolIdx* names = (PoolIdx*)c;
+                for (size_t j = 0; j < i.callFixedArgs.nargs; j++)
+                    names[j] = Pool::insert(ReadItem(refTable, inp));
+            }
+            break;
+        }
+        case Opcode::call_builtin_:
+            i.callBuiltinFixedArgs.nargs = InInteger(inp);
+            i.callBuiltinFixedArgs.ast = Pool::insert(ReadItem(refTable, inp));
+            i.callBuiltinFixedArgs.builtin =
+                Pool::insert(ReadItem(refTable, inp));
+            break;
+        case Opcode::static_call_:
+            i.callFixedArgs.nargs = InInteger(inp);
+            i.callFixedArgs.ast = Pool::insert(ReadItem(refTable, inp));
+            InBytes(inp, &i.callFixedArgs.given, sizeof(Assumptions));
+            i.staticCallFixedArgs.targetClosure =
+                Pool::insert(ReadItem(refTable, inp));
+            i.staticCallFixedArgs.versionHint =
+                Pool::insert(ReadItem(refTable, inp));
+            break;
+        case Opcode::deopt_: {
+            DeoptMetadata* meta =
+                DeoptMetadata::deserialize(code, refTable, inp);
+            size_t size =
+                sizeof(DeoptMetadata) + meta->numFrames * sizeof(FrameInfo);
+            SEXP store = Rf_allocVector(RAWSXP, size);
+            memcpy(DATAPTR(store), meta, size);
```

Comment:
I was thinking, could you do this without the copying? Something like have the `deserialize` function return the already finished sexp? Then you would get rid of the `malloc`ing and `free`ing completely

Reply:
I tried before but it crashes. The difficulty is that the R GC can run while the code body is deserialized, but if the code is in an SEXP the GC will try to look through it, including looking through the body. I'll try to set the gc length to 0

Reply:
I forgot I was logged in the wrong account for the above comment.